x/programs/rust/examples/csamm/src/lib.rs

```diff
@@ -0,0 +1,169 @@
+use std::cmp::Ordering;
+use wasmlanche_sdk::{params, program::Program, public, state_keys, types::Address};
+
+// Implementation of a CSAMM. This assumes a 1:1 peg between the two assets or one of them is going to be missing from the pool.
+
+/// The program state keys.
+#[state_keys]
+enum StateKeys {
+    // TODO this is also an LP token, could we reuse parts of the "token" program ?
+    // TotalSupply,
+    // Balance(Address),
+    /// The vec of coins. Key prefix 0x0 + addresses
+    // TODO Vec doesn't implement Copy, so can we not define dynamic array in the program storage ?
+    // Coins([Address; 2]),
+    CoinX,
+    CoinY,
+}
+
+/// Initializes the program address a count of 0.
+#[public]
+fn constructor(program: Program, coin_x: Program, coin_y: Program) -> bool {
+    if program.state().get::<Program, _>(StateKeys::CoinX).is_ok()
```

Comment:
Interesting, how do you define initialization for a program ? The name `constructor` was chosen arbitrarily because I couldn't see a specific pattern for initialization, maybe should I rename it to `init` like in the token example ?